CMakeLists.txt

```diff
@@ -521,6 +525,41 @@ if(EMSCRIPTEN)
   target_link_libraries(binaryen_js debug "--profiling")
   target_link_libraries(binaryen_js debug "-sASSERTIONS")
   install(TARGETS binaryen_js DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+  # binaryen.ts TypeScript variant (leverages the wasm variant)
+  # compile the typescript file that wraps the wasm
+  add_custom_target(binaryen_ts_wrapper
+              COMMAND tsc --target ES2020 --module ES2022 --declaration --declarationMap
+                      ${CMAKE_CURRENT_SOURCE_DIR}/src/ts/binaryen_ts.ts &&
+                      mv ${CMAKE_CURRENT_SOURCE_DIR}/src/ts/binaryen_ts.js ${CMAKE_BINARY_DIR}/bin/binaryen_ts.js &&
+                      mv ${CMAKE_CURRENT_SOURCE_DIR}/src/ts/binaryen_ts.d.ts ${CMAKE_BINARY_DIR}/bin/binaryen_ts.d.ts &&
+                      mv ${CMAKE_CURRENT_SOURCE_DIR}/src/ts/binaryen_ts.d.ts.map ${CMAKE_BINARY_DIR}/bin/binaryen_ts.d.ts.map)
+  add_custom_target(binaryen_post_ts
+              COMMAND tsc --target ES2020 --module ES2022
+                      ${CMAKE_CURRENT_SOURCE_DIR}/src/ts/binaryen-post.ts)
+  add_executable(binaryen_wasm_ts
+                 ${binaryen_emscripten_SOURCES})
+  add_dependencies(binaryen_wasm_ts binaryen_ts_wrapper binaryen_post_ts)
+  target_link_libraries(binaryen_wasm_ts wasm asmjs emscripten-optimizer passes ir cfg support analysis parser wasm)
+  target_link_libraries(binaryen_wasm_ts "-sFILESYSTEM")
+  target_link_libraries(binaryen_wasm_ts "-sEXPORT_NAME=Binaryen")
+  target_link_libraries(binaryen_wasm_ts "-sNODERAWFS=0")
+  target_link_libraries(binaryen_wasm_ts "-sEXPORT_ES6")
+  target_link_libraries(binaryen_wasm_ts "-sEXPORTED_RUNTIME_METHODS=allocateUTF8OnStack,stringToAscii")
+  target_link_libraries(binaryen_wasm_ts "-sEXPORTED_FUNCTIONS=_malloc,_free")
```

Comment:
Is it possible to avoid having to repeat all these flags from the vanilla binaryen_wasm build? Do we need another target at all? Can we unconditionally emit the TS wrapper for all binaryen_js and binaryen_wasm builds instead?

Reply:
An option I haven't considered is to apply for JS the same approach than TS i.e. rather than embed the JS code in the generated JS, have it live aside and consume the wasm. That way we would have a single wasm target generating:
 - binaryen_wasm.js
 - binaryen_js.js
 - binaryen_ts.js
 _ binaryen_ts.d.ts

Reply:
I actually tried out this approach as an experiment, and technically speaking it would work. Let me know if that's a better idea and I'll evolve this PR accordingly.

That said it not only impacts the binaryen_wasm.js target, but also the binaryen_js.js one, since they share the binaryen.js-post.js post-generation file.

Reply:
@kripken, what do you think about this?

Reply:
Regarding adding another file here, that seems possible in general but we do build for node.js using `-sSINGLE_FILE`. I'm actually not sure why we do that, but if we want to add another file here we should probably look into removing `SINGLE_FILE` first.

Separately, could another option be to always build for TS and let users use a TS-to-JS transpiler if they don't want TS? How is this kind of thing usually done in the JS/TS ecosystem?

Reply:
The reason I'm keeping the current JS is backwards compatibility. The build script already creates a JS from TS, but it's a bit different from the current one. If binaryen_wasm only has a few consumers then it's obviously simpler to drop the current one and advertise the breaking changes.

Reply:
And it's done the usual way I.e. run tsc, which generates the JS file, a d.ts file and a .map file for debugging. Pretty standard.

Reply:
@kripken can you point me to where we is the -sSINGLE_FILE option ?

Reply:
I see, thanks.

When you say "a bit different" is that stuff like code size or does it also behave differently? Is it a drop-in replacement?

Reply:
The new file behaves differently:

1. the current file returns a function which must be awaited in order to access the various entry points, whereas the new one directly exposes the entry points
2. a few entry points have been moved or renamed for consistency
3. xxxID constants have been dropped, purely out of lazyness - I can't think of a scenario where one would need them

I can't change 1, but I can fix 2 and 3 if 1 is acceptable and that leads do a single source of truth and just 2 targets (like we have today) instead of 3.

Reply:
I tried turning off SINGLE_FILE. The purpose of this flag seems to be the following:
 - the first generated file is a .wasm file 
 - if SINGLE_FILE is off, a .js file is generated, with code to consume the .wasm data, but locating the file is left to the application
 - if SINGLE_FILE is on, the binary .wasm is serialized to base64 and embedded in the .js
(that doesn't affect the async loading requirement which is the blocking point for typescript, because you can't declare const and enums at runtime)